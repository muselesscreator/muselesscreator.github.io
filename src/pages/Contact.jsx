import 'react';
import { useKeyedState } from '@muselesscreator/use-keyed-state';
import { StrictDict } from '@muselesscreator/strict-dict';

const stateKeys = StrictDict({
  fullname: 'fullname',
  email: 'email',
  message: 'message',
});

const About = () => {
  const [fullname, setFullname] = useKeyedState(stateKeys.fullname, '');
  const [email, setEmail] = useKeyedState(stateKeys.email, '');
  const [message, setMessage] = useKeyedState(stateKeys.message, '');

  const isValid = fullname && email && message;

  return (
    <article className="contact active">
      <header>
        <h2 className="h2 article-title">Contact</h2>
      </header>
      <section className="contact-form">
        <h3 className="h3 form-title">Contact Form</h3>
        <form method="POST" action="https://formspree.io/f/mkndboev" className="form" data-form>
          <div className="input-wrapper">
            <input
              type="text"
              name="fullname"
              className="form-input"
              placeholder="Full name"
              required
              value={fullname}
              onChange={(e) => setFullname(e.target.value)}
            />
            <input
              type="email"
              name="email"
              className="form-input"
              placeholder="Email address"
              required
              value={email}
              onChange={(e) => setEmail(e.target.value)}
            />
          </div>
          <textarea
            name="message"
            className="form-input"
            placeholder="Your Message"
            required
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          ></textarea>
          <button className="form-btn" type="submit" disabled={!isValid}>
            <ion-icon name="paper-plane"></ion-icon>
            <span>Send Message</span>
          </button>
        </form>
      </section>
    </article>
  );
};

export default About;
