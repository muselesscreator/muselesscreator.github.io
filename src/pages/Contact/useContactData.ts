import { useKeyedState } from '@muselesscreator/use-keyed-state';
import { StrictDict } from '@muselesscreator/strict-dict';

export const stateKeys = StrictDict({
  fullname: 'fullname',
  email: 'email',
  message: 'message',
}) as Record<string, string>;

export type ContactData = {
  fullname: string,
  email: string,
  message: string,
  isValid: boolean,
  handleNameChange: (e: { target: { value: string } }) => void,
  handleEmailChange: (e: { target: { value: string } }) => void,
  handleMessageChange: (e: { target: { value: string } }) => void,
};

const useContactData = (): ContactData => {
  const [fullname, setFullname] = useKeyedState<string>(stateKeys.fullname, '');
  const [email, setEmail] = useKeyedState<string>(stateKeys.email, '');
  const [message, setMessage] = useKeyedState<string>(stateKeys.message, '');

  const isValid = !!(fullname && email && message);

  return {
    fullname,
    email,
    message,
    isValid,
    handleNameChange: (e: { target: { value: string } }) => setFullname(e.target.value),
    handleEmailChange: (e: { target: { value: string } }) => setEmail(e.target.value),
    handleMessageChange: (e: { target: { value: string } }) => setMessage(e.target.value),
  }
};

export default useContactData;
