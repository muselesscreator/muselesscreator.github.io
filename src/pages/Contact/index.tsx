import SendIcon from '@mui/icons-material/Send';

import useContactData from './useContactData';

export const formAction = 'https://formspree.io/f/mkndboev';

const Contact = () => {
  const {
    fullname,
    email,
    message,
    isValid,
    handleNameChange,
    handleEmailChange,
    handleMessageChange,
  } = useContactData();

  const inputs = [
    {
      type: 'text',
      name: 'fullname',
      placeholder: 'Full name',
      value: fullname,
      onChange: handleNameChange,
    },
    {
      type: 'email',
      name: 'email',
      placeholder: 'Email address',
      value: email,
      onChange: handleEmailChange,
    },
  ];

  return (
    <article className="contact active">
      <header>
        <h2 className="h2 article-title">Contact</h2>
      </header>
      <section className="contact-form">
        <h3 className="h3 form-title">Contact Form</h3>
        <form method="POST" action={formAction} className="form" data-form>
          <div className="input-wrapper">
            {...inputs.map((input) => (
              <input
                key={input.name}
                type={input.type}
                name={input.name}
                className="form-input"
                placeholder={input.placeholder}
                required
                value={input.value as string}
                onChange={input.onChange}
              />
            ))}
          </div>
          <textarea
            name="message"
            className="form-input"
            placeholder="Your Message"
            required
            value={message as string}
            onChange={handleMessageChange}
          ></textarea>
          <button className="form-btn" type="submit" disabled={!isValid}>
            <SendIcon />
            <span>Send Message</span>
          </button>
        </form>
      </section>
    </article>
  );
};

export default Contact;
