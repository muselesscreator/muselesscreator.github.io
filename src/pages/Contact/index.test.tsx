import {
  vi,
  describe,
  it,
  test,
  expect,
  Mock,
} from 'vitest';
import { shallow, setMockName } from '@muselesscreator/react-shallow-snapshot';

import useContactData, from './useContactData';

import Contact from '.';

vi.mock('./useContactData', () => ({ default: vi.fn() }));
vi.mock('@mui/icons-material/Send', () => ({ default: 'SendIcon' }));

const hookProps = {
  fullname: 'Test Full Name',
  email: 'testEmail@google.com',
  message: 'This is a message for testing purposes',
  isValid: true,
  handleNameChange: vi.fn(),
  handleEmailChange: vi.fn(),
  handleMessageChange: vi.fn(),
};
setMockName(hookProps.handleNameChange, 'hookProps.handleNameChange');
setMockName(hookProps.handleEmailChange, 'hookProps.handleEmailChange');
setMockName(hookProps.handleMessageChange, 'hookProps.handleMessageChange');

const invalidHookProps = {
  ...hookProps,
  isValid: false,
  fullname: '',
  email: '',
  message: '',
};
(useContactData as Mock).mockReturnValue(hookProps);

describe('Contact page', () => { 
  describe('behavior', () => { 
    it('loads contact data from hook', () => {
      shallow(<Contact />);
      expect(useContactData).toHaveBeenCalled();
    });
  });
  describe('output', () => {
    test('snapshot', () => {
      const el = shallow(<Contact />);
      expect(el.snapshot).toMatchSnapshot();
    });
    it('provides inputs for name and email, and a textarea for a message', () => {
      const el = shallow(<Contact />);
      const inputProps = el.instance.findByType('input')
        .map((input: { props: Record<string, string> }) => input.props);
      const textareaProps = el.instance.findByType('textarea')[0].props;
      expect(inputProps[0].value).toBe(hookProps.fullname)
      expect(inputProps[1].value).toBe(hookProps.email)
      expect(textareaProps.value).toBe(hookProps.message)
      expect(inputProps[0].onChange).toBe(hookProps.handleNameChange)
      expect(inputProps[1].onChange).toBe(hookProps.handleEmailChange)
      expect(textareaProps.onChange).toBe(hookProps.handleMessageChange)
      expect(inputProps[0].type).toBe('text');
      expect(inputProps[1].type).toBe('email');
    });
    test('invalid (empty) snapshot (disables submit button', () => {
      (useContactData as Mock).mockReturnValue(invalidHookProps);
      const el = shallow(<Contact />);
      expect(el.snapshot).toMatchSnapshot();
      expect(el.instance.findByType('button')[0].props.disabled).toBe(true);
      const inputProps = el.instance.findByType('input')
        .map((input: { props: Record<string, string> }) => input.props);
      const textareaProps = el.instance.findByType('textarea')[0].props;
      expect(inputProps[0].value).toBe('')
      expect(inputProps[1].value).toBe('')
      expect(textareaProps.value).toBe('')
      expect(inputProps[0].onChange).toBe(hookProps.handleNameChange)
      expect(inputProps[1].onChange).toBe(hookProps.handleEmailChange)
      expect(textareaProps.onChange).toBe(hookProps.handleMessageChange)
    });
  });
});
