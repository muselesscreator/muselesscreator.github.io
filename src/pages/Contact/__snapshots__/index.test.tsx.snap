// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Contact page > output > invalid (empty) snapshot (disables submit button 1`] = `
<article
  className="contact active"
>
  <header>
    <h2
      className="h2 article-title"
    >
      Contact
    </h2>
  </header>
  <section
    className="contact-form"
  >
    <h3
      className="h3 form-title"
    >
      Contact Form
    </h3>
    <form
      action="https://formspree.io/f/mkndboev"
      className="form"
      data-form={true}
      method="POST"
    >
      <div
        className="input-wrapper"
      >
        <input
          className="form-input"
          name="fullname"
          onChange={[MockFunction hookProps.handleNameChange]}
          placeholder="Full name"
          required={true}
          type="text"
          value=""
        />
        <input
          className="form-input"
          name="email"
          onChange={[MockFunction hookProps.handleEmailChange]}
          placeholder="Email address"
          required={true}
          type="email"
          value=""
        />
      </div>
      <textarea
        className="form-input"
        name="message"
        onChange={[MockFunction hookProps.handleMessageChange]}
        placeholder="Your Message"
        required={true}
        value=""
      />
      <button
        className="form-btn"
        disabled={true}
        type="submit"
      >
        <SendIcon />
        <span>
          Send Message
        </span>
      </button>
    </form>
  </section>
</article>
`;

exports[`Contact page > output > snapshot 1`] = `
<article
  className="contact active"
>
  <header>
    <h2
      className="h2 article-title"
    >
      Contact
    </h2>
  </header>
  <section
    className="contact-form"
  >
    <h3
      className="h3 form-title"
    >
      Contact Form
    </h3>
    <form
      action="https://formspree.io/f/mkndboev"
      className="form"
      data-form={true}
      method="POST"
    >
      <div
        className="input-wrapper"
      >
        <input
          className="form-input"
          name="fullname"
          onChange={[MockFunction hookProps.handleNameChange]}
          placeholder="Full name"
          required={true}
          type="text"
          value="Test Full Name"
        />
        <input
          className="form-input"
          name="email"
          onChange={[MockFunction hookProps.handleEmailChange]}
          placeholder="Email address"
          required={true}
          type="email"
          value="testEmail@google.com"
        />
      </div>
      <textarea
        className="form-input"
        name="message"
        onChange={[MockFunction hookProps.handleMessageChange]}
        placeholder="Your Message"
        required={true}
        value="This is a message for testing purposes"
      />
      <button
        className="form-btn"
        disabled={false}
        type="submit"
      >
        <SendIcon />
        <span>
          Send Message
        </span>
      </button>
    </form>
  </section>
</article>
`;
